TCD-2905:Hydra (#2264)
saucelabs/sauce-docs: docs/error-reporting/advanced/hydra/usage.md

Let's dive deeper into how Hydra can help you eliminate the complex bugs that impact your production applications.

## Overview

### Running Hydra

First things first: how do you run Hydra?

```shell
$ hydra
```

If you have configured coroner, you can also use:

```shell
$ coroner view
```

Running these commands instantly displays the exact root cause of the crash and, in some cases, even proposes a fix for you.

What you actually see is a colorful ncurses interface (unless you pass the `-m` flag to hydra, in which case it is displayed in black and white). The interface reminds you of some of your favorite tools, such as htop and tig. So, how is this interface organized?

### Understanding the Panes

The interface is divided into several panes, each representing a different aspect of the application's state.

The panes, in order, are:

- Threads
- Frames
- Variables
- Router

We'll explain the Router pane in more detail later. For now, know that it contains various metadata about the application, including system information (memory/CPU usage), registers, kernel frames, and more. It also offers configurable integrations, such as source code management.

> Among the panes, the only one that remains constant is the Threads pane (well, almost constant, because you can still change how threads are displayed). The Frames, Variables, and Router panes dynamically update based on the selected context. We refer to these panes as context-aware. Frames are populated based on the presently selected thread, variables and registers depend on the current frame, and so on.

Author: ```suggestion
Among the panes, the only one that remains constant is the Threads pane (well, almost constant because you can still change how threads are displayed). The Frames, Variables, and Router panes dynamically update based on the selected context. We refer to these panes as context-aware. Frames are populated based on the selected thread, Variables and Registers depend on the current frame, and more.
```

crash_app.map=object:^libyajl[\.-],/home/user/projects/yajl
crash_app.ignore=object:^lib
crash_app.map=/home/user/projects/crash_app
crash_app.trigger=/home/user/projects/crash_app,version,git -C %s checkout -q %0
```

The `.map` commands associate an object file or function name with a source code folder for the application. After the colon, you specify a regular expression to match the object or function name, followed by the corresponding source code path. The commands are processed from top to bottom, and the first match determines the path where Hydra searches for the source code.

A `.map` command without a regular expression match acts as a wildcard, matching anything in the same directory.

When using `.ignore=` followed by a regular expression, Hydra ignores matches for subsequent lines. In the example above, object files that start with "lib" and haven't matched any earlier `.map=object` lines are ignored, preventing Hydra from associating them with source code.

### Triggers

You can also trigger a command to run with `.trigger=`. The most common use case is triggering a git checkout of the correct branch when using Hydra.

> To use a trigger, specify the source code path as the first parameter and the command to execute as the last parameter. Between them, you can include key-value pairs that act as positional variables in the trigger command. In the example above, the `version` key-value maps to the `%0` parameter. Additional key-value pairs correspond to `%1`, `%2`, and so on. `%s` refers to the project path.

Author: ```suggestion
To use a trigger, specify the source code path as the first parameter and the command to run as the last parameter. Between them, you can include key-value pairs that act as positional variables in the trigger command. In the example above, the `version` key-value maps to the `%0` parameter. Additional key-value pairs correspond to `%1`, `%2`, and so on. `%s` refers to the project path.
```

#### Troubleshooting Triggers

If a trigger you've set up doesn't seem to be firing, keep in mind the following:

- A trigger fires only when code from the specified folder is accessed by Hydra, which occurs when a frame using that code is highlighted.
- A trigger fires only if:
  - The path in the trigger line was previously matched by a `.map=` line.
  - At least one key-value attribute is specified in the `.trigger=` line.
  - The listed attributes are present in the snapshot (that is, `ptrace` was called with the `--kv=` flag and included the specified attributes).

### Item Collapsing

Context: Any pane displaying a list with a hierarchy indicated by `+/-` symbols

Commands:

- Collapse or expand an item: +/- symbols next to the item
- Collapse or expand all items: Commands specific to each pane

> Any item with an indicated hierarchy (that is, a +/- symbol) can be collapsed or expanded to hide or show its "children." In the Threads pane, children may refer to members of a specific thread group; in the Variables pane, members of a struct or array; in the Process pane, structured heap metadata (arenas, thread caches, etc.); and so on.

Author: ```suggestion
Any item with an indicated hierarchy (such as a +/- symbol) can be collapsed or expanded to hide or show its "children." In the Threads pane, children may refer to members of a specific thread group; in the Variables pane, members of a struct or array; in the Process pane, structured heap metadata (arenas, thread caches, etc.); and more.
```

There is one exception to the default collapsing behavior: inlined variable annotations. See the [Inlined Annotations](#inlined-annotations) section for details.

### Inlined Annotations

Any annotations on a variable are displayed directly below the variable. If a variable chain collapses, but one of the variables in the chain is annotated, the minimum number of variables necessary (including the annotated variable and its owners) is displayed along with the annotation itself.

### Annotation Jumping

> Variables across frames are annotatable, and in a single frame, there may be thousands of variables, making it challenging to see any annotations. Annotation jumping comes in handy here. Open the `Warnings` tab in the bottom pane (press `w`), scroll through the list to find the annotation you're interested in, and press **Enter**. The thread, frame, and variable views update to the position of the annotation's owner.

Author: ```suggestion
Variables across frames are annotatable, and in a single frame, there may be thousands of variables, making it challenging to see any annotations. Annotation jumping proves valuable in this case. Open the `Warnings` tab in the bottom pane (press `w`), scroll through the list to find the annotation you're interested in, and press **Enter**. The thread, frame, and variable views update to the position of the annotation's owner.
```

### Pane Maximization

Context: Any pane

Commands:

- Maximize or restore panes: `M (Shift + m)`

> All panes support maximization. When a pane is maximized, it has an associated context (for example, a maximized Thread pane has a Frame pane context). The maximized pane occupies the majority of the screen space, while the contextual panes occupy the rest. All other panes are hidden. To restore all panes and sizes, you can press `M` again or use a macro movement hotkey to switch to one of the hidden panes (moving between shown panes does not restore the sizes).

Author: ```suggestion
All panes support maximization. When a pane is maximized, it has an associated context (for example, a maximized Thread pane has a Frame pane context). The maximized pane occupies most of the screen space, while the contextual panes occupy the rest. All other panes are hidden. To restore all panes and sizes, press `M` again or use a macro movement hotkey to switch to one of the hidden panes (moving between shown panes does not restore the sizes).
```

> ### Regex Search

Author: ```suggestion
### Regula Expression Search
```

Context: Any pane displaying a list

Commands:

- Search: `/`
- Go to the next search result: `n`
- Go to the previous search result: `N`

> All panes displaying a list support regex searches. Each column is searched independently. For example, in the Threads pane, the search applies to status, TID, basename, thread name, and top frame symbol.

Author: ```suggestion
All panes displaying a list support regular expression searches. Each column is searched independently. For example, in the Threads pane, the search applies to status, TID, basename, thread name, and top frame symbol.
```

### Index Jumping

Context: Any pane

Commands:

- Jump to a specific index: `:`

All panes support index jumping. Except for the Source Code Management (SCM) pane, the indices are 0-based. The SCM pane uses a 1-based index (similar to a vim file). If an index is outside the valid range, it jumps to the first or last elements, respectively.

### Thread Grouping

Context: Any pane for commands, Thread pane for display

Commands:

- Group threads: `:group`
- Supported group types: `callstack`
- Ungroup threads: `:ungroup`
- Sort threads: `:sort [sort-type]`
- Supported sort types: `tid`
- Reverse sorting order: `:rsort`

Threads are automatically grouped based on the selected group type and sorted according to the specified sort type.

> With callstack grouping, threads with identical callstacks are grouped together. Using the tid sort type orders threads based on their thread IDs. There are no other supported group and sort types.

Author: ```suggestion
With callstack grouping, threads with identical callstacks are grouped together. Using the tid sort type orders threads based on their thread IDs. No other supported group and sort types are available.
```

- `l`: Scroll right (if there is text off the screen)
- `H`: Move to the top of the current view (vim behavior)
- `L`: Move to the bottom of the current view
- `PgUp`: Move up one full page of items
- `PgDn`: Move down one full page of items
- `Home`: Move to the first item
- `End`: Move to the last item
- `:` Jump to the specified position
- `tab`: Move to the next pane
- `1-4`: Move to the pane associated with the number
  - `1`: Threads
  - `2`: Frames
  - `3`: Variables
  - `4`: Router

### State Jumping/Linking

Context: Any non-router pane

Commands:

- `u`: Show the position of the current selection in the current pane
- `:j`: Refocus to the provided position

Any state in the top three panes (Threads, Frames, and Variables) can be immediately refocused by "jumping" to its position, similar to annotation jumping. Press `u` to show the position URL of the current selection and use the global `:j` command to refocus on that state.

### Global Commands

- `:s`
- `:!`
- `:j`
- `:`
- `/`
- `:sort`
- `:rsort`
- `:group`
- `:ungroup`
- `:q`

#### Immediately Run Global Commands

Context: Hydra command-line parameter

Commands:

- -e "command"

> All global commands, except regex search, can be executed immediately upon starting Hydra. This is beneficial for sharing state with other users. Provide them with a snapshot and a position URL (using the `u` command) and have them open it using: `hydra -e "j <position URL>"`.

Author: ```suggestion
All global commands, except regular expression search, can be executed immediately upon starting Hydra. This is beneficial for sharing the state with other users. Provide them with a snapshot and a position URL (using the `u` command) and have them open it using: `hydra -e "j <position URL>"`.
```

### Global/TLS Variables

Displays the variables with global and thread-local storage that were requested at the time of the trace using `ptrace` (refer to the `ptrace` documentation for instructions). Variables are organized into a hierarchy of `[Thread (for TLS variables)] - [Object] - [Compilation Unit (CU)]`.

### Source Code Integration

Commands:

- Open a source file in the configured editor
- Center view on the last-executed line of the frame

Shows the source code for the presently selected frame. Index jumping is supported, but regex searches are not. The initial line selected is the last-executed line of the frame.

### Annotations

Commands:

- Jump to the annotation owner (Hydra refocuses on the thread, frame, or variable owning the annotation; Backtrace annotations are not jumpable.)

Lists all annotations contained in the snapshot, excluding JSON-type annotations. You can jump to a selected annotation by pressing **Enter**.

JSON-type annotations are shown in either the Process or Context router tabs. Refer to those sections for more details.

### Column Specification

> The following is the column specification for each pane, from left to right (panes with a single column or containing simple key-value lists are omitted):

Author: ```suggestion
The following is the column specification for each pane, from left to right (panes with a single column or containing key-value lists are omitted):
```
